{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reverse complement if needed, /rc previously added from bam\n",
    "import sys\n",
    "from Bio import SeqIO\n",
    "in_fname = sys.argv[1]\n",
    "out_fname = sys.argv[2]\n",
    "\n",
    "complement = {'A':'T', 'C':'G', 'G':'C', 'T':'A'}\n",
    "with open(in_fname) as f, open(out_fname, 'w') as out_f:\n",
    "    for hap in SeqIO.parse(f, format=\"fasta\"):\n",
    "        out_seqname = hap.name.split(':')[0].replace('/rc', '').replace('h1', 'hap1').replace('h2', 'hap2')\n",
    "        out_f.write(f\">{out_seqname}\\n\")\n",
    "        if '/rc' in hap.name:\n",
    "            out_f.write(''.join([complement[z] for z in hap.seq[::-1]]) + '\\n')\n",
    "        else:\n",
    "            out_f.write(str(hap.seq) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out haplotypes that map to multiple contigs/ multiple places on contig\n",
    "from Bio import SeqIO\n",
    "from collections import defaultdict\n",
    "import sys\n",
    "\n",
    "in_fname = sys.argv[1]\n",
    "out_fname_no_multiply_mapping = sys.argv[2]\n",
    "\n",
    "seqs_per_hap = defaultdict(list)\n",
    "with open(in_fname) as f:\n",
    "    for hap in SeqIO.parse(f, format=\"fasta\"):\n",
    "        seqs_per_hap[hap.name.split(':')[0].replace('/rc', '')].append(str(hap.seq))\n",
    "\n",
    "with open(out_fname_no_multiply_mapping, 'w') as out_f:\n",
    "    for hap in sorted(seqs_per_hap):\n",
    "        if len(set(seqs_per_hap[hap])) == 1:\n",
    "            out_f.write(f'>{hap}\\n{seqs_per_hap[hap][0]}\\n')\n",
    "        else:\n",
    "            pass #print('Skipped ' + hap)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#=======Create k-mer plot with colored blocks for CYP2D6 and CYP2D7==========="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!git clone https://github.com/jwertz01/KmerShade.git\n",
    "!cd KmerShade; snakemake --cores 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#=======Assign star alleles==========="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Align haplotype sequences to reverse-complemented star allele reference sequences from PharmVar\n",
    "!wget -O CYP2D6-6.2.1_pharmvar.zip \"https://www.pharmvar.org/get-download-file?name=CYP2D6&refSeq=ALL&fileType=zip&version=current\"\n",
    "!minimap2 -x asm20 -a -Y --MD --eqx --sam-hit-only -N 2000 all_contigs_concat_revcomp_no_multiply_mapping.fa CYP2D6-6.2.1_pharmvar/CYP2D6.haplotypes.revcomp.fa > cyp_seqs_aligned_star_alleles.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert and index\n",
    "!samtools view -b cyp_seqs_aligned_star_alleles.sam | samtools sort > cyp_seqs_aligned_star_alleles.bam\n",
    "!samtools index cyp_seqs_aligned_star_alleles.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Get star allele site matches via CIGAR string parsing of reference star\n",
    "allele vs. haplotype alignments'''\n",
    "import argparse\n",
    "import sys\n",
    "import os\n",
    "from collections import defaultdict, namedtuple\n",
    "import pysam\n",
    "from pathlib import Path\n",
    "\n",
    "VcfRec = namedtuple('VcfRec', ['pos', 'ref', 'alt'])\n",
    "\n",
    "def main():\n",
    "    args = parse_args()\n",
    "    assert args.pharmvar_dir.is_dir()\n",
    "    sites_per_star_allele = load_star_allele_definitions(\n",
    "        args.pharmvar_dir, args.seq_start_pos\n",
    "    )\n",
    "    matched_sites = defaultdict(lambda: defaultdict(set))\n",
    "\n",
    "    sam = args.sam\n",
    "    align_f = pysam.AlignmentFile(sam, 'rb') if (\n",
    "        sam.suffix == '.bam'\n",
    "    ) else pysam.AlignmentFile(sam)\n",
    "    for i, r in enumerate(align_f):\n",
    "        if i % 10000 == 0:\n",
    "            print(i)\n",
    "        process_alignment(\n",
    "            r, sites_per_star_allele, matched_sites, args.exclude_haps,\n",
    "            args.indel_buffer\n",
    "        )\n",
    "\n",
    "    write_output(matched_sites, args.out)\n",
    "\n",
    "\n",
    "def parse_args():\n",
    "    '''Parse command line arguments'''\n",
    "    p = argparse.ArgumentParser(\n",
    "        description='Identifies star allele site matches from reference star '\n",
    "        'allele vs. haplotype alignments'\n",
    "    )\n",
    "    p.add_argument(\n",
    "        'sam', type=Path, help='SAM file with star allele reference sequences '\n",
    "        '(query) aligned to haplotypes (reference). Please use minimap2 '\n",
    "        'with --eqx and with a high -N.'\n",
    "    )\n",
    "    p.add_argument(\n",
    "        'pharmvar_dir', type=Path, help='Directory with PharmVar VCF files'\n",
    "    )\n",
    "    p.add_argument('out', type=Path, help='Final star allele TSV')\n",
    "    p.add_argument(\n",
    "        '--seq_start_pos', type=int, default=42124499, \n",
    "        help='Starting position of star allele sequences on the chromosome '\n",
    "        '(42124499)'\n",
    "    )\n",
    "    p.add_argument(\n",
    "        '--exclude_haps', nargs='+',\n",
    "        help='List of haplotypes to exclude, e.g. SV haplotypes (none)',\n",
    "        default=[]\n",
    "    )\n",
    "    p.add_argument(\n",
    "        '--indel_buffer', type=int,\n",
    "        help='Number of surrounding bases that must match between star allele '\n",
    "        'and haplotype to count as a site match for indels (20).',\n",
    "        default=20\n",
    "    )\n",
    "    if len(sys.argv) < 4:\n",
    "        p.print_help()\n",
    "        sys.exit(0)\n",
    "    return p.parse_args()\n",
    "\n",
    "\n",
    "def load_star_allele_definitions(vcf_dir, seq_start_pos):\n",
    "    '''Load star allele definitions from VCF files.'''\n",
    "    sites_per_star_allele = defaultdict(list)\n",
    "    for fname in vcf_dir.glob('*.vcf'):\n",
    "        star_allele = fname.stem\n",
    "        offset = 0\n",
    "        with open(fname) as f:\n",
    "            for line in f:\n",
    "                if not line.strip('\"').startswith('#'):\n",
    "                    chrom, pos, _, ref, alt, *_ = line.strip().split('\\t')\n",
    "                    # Position of variant from start of reference star allele\n",
    "                    pos = int(pos) - seq_start_pos + offset\n",
    "                    sites_per_star_allele[star_allele].append(\n",
    "                        VcfRec(pos, ref, alt)\n",
    "                    )\n",
    "                    # Account for position shifts due to indels\n",
    "                    offset += len(alt) - len(ref)\n",
    "    return sites_per_star_allele\n",
    "\n",
    "\n",
    "def process_alignment(\n",
    "    r, sites_per_star_allele, matched_sites, exclude_haps, indel_buffer\n",
    "):\n",
    "    '''Process a single alignment record.'''\n",
    "    star_allele = r.query_name.replace('*', '_')\n",
    "    hap = r.reference_name\n",
    "    if hap in exclude_haps:\n",
    "        return\n",
    "    star_allele_sites = sites_per_star_allele[star_allele]\n",
    "    sites_pos = [z.pos for z in star_allele_sites]\n",
    "    star_to_hap_pos = {}\n",
    "    start_hap_pos = None\n",
    "    for star_pos, hap_pos in r.get_aligned_pairs():  # (ref, hap)\n",
    "        if start_hap_pos is None and hap_pos is not None:\n",
    "            start_hap_pos = hap_pos\n",
    "        if star_pos in sites_pos:\n",
    "            star_to_hap_pos[star_pos] = hap_pos\n",
    "    # Sequences for aligned segments\n",
    "    hap_seq = r.get_reference_sequence().upper()\n",
    "    star_seq = r.query_sequence.upper()\n",
    "    for site in star_allele_sites:\n",
    "        hap_pos = star_to_hap_pos[site.pos]\n",
    "        if hap_pos is None:\n",
    "            continue\n",
    "        # Pos from start of alignment rather than hap seq\n",
    "        hap_pos -= start_hap_pos\n",
    "        match_found = check_site_match(\n",
    "            site, star_seq, hap_pos, hap_seq, indel_buffer, hap, star_allele\n",
    "        )\n",
    "        if match_found:\n",
    "            matched_sites[hap][star_allele].add(site)\n",
    "\n",
    "\n",
    "def check_site_match(\n",
    "    site, star_seq, hap_pos, hap_seq, buffer, hap, star_allele\n",
    "):\n",
    "    '''Check if a site matches between star allele and haplotype.'''\n",
    "    hap_alt = hap_seq[hap_pos:hap_pos + len(site.alt)]\n",
    "    if len(site.ref) == len(site.alt):\n",
    "        return hap_alt == site.alt\n",
    "    else:  # Handle indels\n",
    "        surrounding_hap = hap_seq[\n",
    "            max(0, hap_pos - buffer):hap_pos + len(site.alt) + buffer\n",
    "        ]\n",
    "        surrounding_star = star_seq[\n",
    "            max(0, site.pos - buffer):site.pos + len(site.alt) + buffer\n",
    "        ]\n",
    "        return surrounding_hap == surrounding_star\n",
    "\n",
    "\n",
    "def write_output(matched_sites, out_file):\n",
    "    '''Write matched sites to the output file.'''\n",
    "    with open(out_file, 'w') as out_f:\n",
    "        out_f.write('Sample_hap\\tStar_allele\\tMatched_sites\\n')\n",
    "        for sample_hap, star_alleles in sorted(matched_sites.items()):\n",
    "            for star_allele, sites in sorted(star_alleles.items()):\n",
    "                sites_str = ','.join(\n",
    "                    f\"{s.pos}:{s.ref}>{s.alt}\" for s in sorted(sites)\n",
    "                )\n",
    "                out_f.write(\n",
    "                    f\"{sample_hap}\\t{star_allele.replace('CYP2D6_', '*')}\\t\"\n",
    "                    f\"{sites_str}\\n\"\n",
    "                )\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "'''Determine the best star allele based on site and genomic matches.'''\n",
    "import argparse\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import re\n",
    "import math\n",
    "from collections import defaultdict, namedtuple\n",
    "from pathlib import Path\n",
    "from pysam import *\n",
    "\n",
    "VcfRec = namedtuple('VcfRec', ['pos', 'ref', 'alt'])\n",
    "StarMatch = namedtuple('StarMatch', ['star_allele', 'match', 'total'])\n",
    "\n",
    "def main():\n",
    "    args = parse_args()\n",
    "    assert args.pharmvar_dir.is_dir()\n",
    "    best_genomic_matches = find_best_genomic_matches(\n",
    "        args.sam, args.exclude_haps, args.ignore_alleles\n",
    "    )\n",
    "    sites_per_star_allele = load_star_allele_definitions(\n",
    "        args.pharmvar_dir, args.seq_start_pos\n",
    "    )\n",
    "    all_matched_sites = read_in_all_matched_sites(\n",
    "        args.matched_sites, args.exclude_haps, args.ignore_alleles\n",
    "    )\n",
    "\n",
    "    with open(args.out, 'w') as out_f:\n",
    "        w = csv.DictWriter(out_f, delimiter='\\t', fieldnames=[\n",
    "            'Hap', 'Final star allele', 'Best genomic match', \n",
    "            'Best genomic match %', 'Best site match', 'Best site match %', \n",
    "            'Second-best site match (DIFFERENT major star allele)',\n",
    "            'Second-best site match %',\n",
    "        ])\n",
    "        w.writeheader()\n",
    "        for hap in all_matched_sites:\n",
    "            write_alleles(\n",
    "                w, hap, all_matched_sites[hap], sites_per_star_allele,\n",
    "                [z[0] for z in best_genomic_matches[hap]]\n",
    "            )\n",
    "\n",
    "\n",
    "def parse_args():\n",
    "    '''Parse command line arguments'''\n",
    "    p = argparse.ArgumentParser(\n",
    "        description='Identifies the best star allele per haplotype based on '\n",
    "        'genomic and site matching'\n",
    "    )\n",
    "    p.add_argument(\n",
    "        'sam', type=Path, help='SAM file with star allele reference sequences '\n",
    "        '(query) aligned to haplotypes (reference). Please use minimap2 '\n",
    "        'with --eqx and with a high -N.'\n",
    "    )\n",
    "    p.add_argument(\n",
    "        'matched_sites', type=Path, help='File with matched sites, with header '\n",
    "        'Sample_hap\\tStar_allele\\tMatched_sites, where matched sites '\n",
    "        'is a comma-separated list of pos:ref>alt'\n",
    "    )\n",
    "    p.add_argument(\n",
    "        'pharmvar_dir', type=Path, help='Directory with PharmVar VCF files'\n",
    "    )\n",
    "    p.add_argument('out', type=Path, help='Final star allele TSV')\n",
    "    p.add_argument(\n",
    "        '--seq_start_pos', type=int, default=42124499, \n",
    "        help='Starting position of star allele sequences on the chromosome '\n",
    "        '(42124499)'\n",
    "    )\n",
    "    # exact same sequence as *n.001 or *n.002\n",
    "    star_alleles_w_dups = [\n",
    "        '*1', '*106', '*107', '*108', '*110', '*112', '*113', '*114',\n",
    "        '*115', '*116', '*120', '*122', '*124', '*137', '*140', '*167',\n",
    "        '*169', '*170', '*173', '*174', '*18', '*19', '*2', '*23', '*24',\n",
    "        '*25', '*26', '*27', '*3', '*33', '*34', '*38', '*39', '*4', '*43',\n",
    "        '*44', '*48', '*50', '*62', '*7', '*71', '*75', '*81', '*86', '*89',\n",
    "        '*9', '*91', '*92', '*93', '*96', '*97'\n",
    "    ]\n",
    "    p.add_argument(\n",
    "        '--ignore_alleles', nargs='+',\n",
    "        help=f\"List of star alleles to ignore, e.g. duplicates \"\n",
    "        f\"({', '.join(star_alleles_w_dups)})\",\n",
    "        default=star_alleles_w_dups\n",
    "    )\n",
    "    p.add_argument(\n",
    "        '--exclude_haps', nargs='+',\n",
    "        help='List of haplotypes to exclude, e.g. SV haplotypes (none)',\n",
    "        default=[]\n",
    "    )\n",
    "    if len(sys.argv) < 5:\n",
    "        p.print_help()\n",
    "        sys.exit(0)\n",
    "    return p.parse_args()\n",
    "\n",
    "\n",
    "def find_best_genomic_matches(sam, exclude_haps=[], ignore_alleles=[]):\n",
    "    '''Find best-matching star allele(s) for each haplotype by scoring CIGAR \n",
    "    string over the genomic locus\n",
    "    '''\n",
    "    best_genomic_matches = defaultdict(list)\n",
    "    align_f = AlignmentFile(sam, 'rb') if (\n",
    "        sam.suffix == '.bam'\n",
    "    ) else AlignmentFile(sam)\n",
    "    for r in align_f:\n",
    "        hap = r.reference_name\n",
    "        star_allele = r.query_name.replace('CYP2D6', '')\n",
    "        if hap in exclude_haps or star_allele in ignore_alleles:\n",
    "            continue\n",
    "        # genomic match score = num bases match - (mismatch + del + indel)\n",
    "        score = 0\n",
    "        for op, op_len in r.cigartuples:\n",
    "            if op == CEQUAL:\n",
    "                score += op_len\n",
    "            elif op in [CDIFF, CINS, CDEL]:\n",
    "                score -= op_len\n",
    "        # Reset best matches if star allele is higher-scoring\n",
    "        if (\n",
    "            (hap not in best_genomic_matches) or\n",
    "            (score > best_genomic_matches[hap][0][1])\n",
    "        ):\n",
    "            best_genomic_matches[hap] = [(star_allele, score)]\n",
    "        # Add to best matches if same score\n",
    "        elif score == best_genomic_matches[hap][0][1] and (\n",
    "            (star_allele, score) not in best_genomic_matches[hap]\n",
    "        ):\n",
    "            best_genomic_matches[hap].append((star_allele, score))\n",
    "    return best_genomic_matches\n",
    "\n",
    "\n",
    "def load_star_allele_definitions(vcf_dir, seq_start_pos):\n",
    "    '''Load star allele definitions from VCF files'''\n",
    "    sites_per_star_allele = defaultdict(list)\n",
    "    for fname in vcf_dir.glob('*.vcf'):\n",
    "        star_allele = fname.stem.replace('CYP2D6_', '*')\n",
    "        offset = 0\n",
    "        with open(fname) as f:\n",
    "            for line in f:\n",
    "                if not line.strip('\"').startswith('#'):\n",
    "                    chrom, pos, _, ref, alt, *_ = line.strip().split('\\t')\n",
    "                    # Position of variant from start of reference star allele\n",
    "                    pos = int(pos) - seq_start_pos + offset\n",
    "                    sites_per_star_allele[star_allele].append(\n",
    "                        VcfRec(pos, ref, alt)\n",
    "                    )\n",
    "                    # Account for position shifts due to indels\n",
    "                    offset += len(alt) - len(ref)\n",
    "    return sites_per_star_allele\n",
    "\n",
    "\n",
    "def read_in_all_matched_sites(\n",
    "    matched_sites_fname, exclude_haps=[], ignore_alleles=[]\n",
    "):\n",
    "    '''Read in matching sites between each star allele and hap'''\n",
    "    all_matched_sites = defaultdict(lambda: defaultdict(set))\n",
    "    with open(matched_sites_fname) as f:\n",
    "        r = csv.DictReader(f, delimiter='\\t')\n",
    "        for line in r:\n",
    "            hap = line['Sample_hap']\n",
    "            star_allele = line['Star_allele']\n",
    "            if hap in exclude_haps or star_allele in ignore_alleles:\n",
    "                continue\n",
    "            for site_str in line['Matched_sites'].split(','):\n",
    "                all_matched_sites[hap][star_allele].add(\n",
    "                    VcfRec(int(z[0]), z[1], z[2]) for z in\n",
    "                    re.split(r'[:>]', site_str)\n",
    "                )\n",
    "    return all_matched_sites\n",
    "\n",
    "\n",
    "def write_alleles(\n",
    "    w, hap, all_matched_sites_hap, sites_per_star_allele, best_genomic_star\n",
    "):\n",
    "    '''Write output TSV line combining best genomic match and site match star\n",
    "    alleles for haplotype.'''\n",
    "    site_matches = []\n",
    "    for star_allele in all_matched_sites_hap:\n",
    "        matched_sites = len(all_matched_sites_hap[star_allele])\n",
    "        total_sites = len(sites_per_star_allele[star_allele])\n",
    "        assert matched_sites <= total_sites\n",
    "        site_matches.append(\n",
    "            StarMatch(star_allele, matched_sites, total_sites)\n",
    "        )\n",
    "    (\n",
    "        best_site_percentage, best_site_matches,\n",
    "        second_best_site_percentage, second_best_site_matches\n",
    "    ) = get_best_site_matches(site_matches)\n",
    "\n",
    "    final_star_allele = get_final_star_allele(\n",
    "        best_genomic_star,\n",
    "        [z.star_allele for z in best_site_matches],\n",
    "        best_site_percentage\n",
    "    )\n",
    "    out_d = {\n",
    "        'Hap':hap,\n",
    "        'Final star allele': final_star_allele,\n",
    "        'Best genomic match': ','.join(best_genomic_star),\n",
    "        'Best genomic match %': match_frac_str([\n",
    "            z for z in site_matches if z.star_allele in best_genomic_star\n",
    "        ]),\n",
    "        'Best site match': ','.join([\n",
    "            z.star_allele for z in best_site_matches\n",
    "        ]),\n",
    "        'Best site match %': match_frac_str(best_site_matches),\n",
    "        'Second-best site match (DIFFERENT major star allele)': ','.join([\n",
    "            z.star_allele for z in second_best_site_matches\n",
    "        ]),\n",
    "        'Second-best site match %': match_frac_str(\n",
    "            second_best_site_matches\n",
    "        )\n",
    "    }\n",
    "    w.writerow(out_d)\n",
    "\n",
    "\n",
    "def get_final_star_allele(\n",
    "    best_genomic_stars, best_site_stars, best_site_percentage\n",
    "):\n",
    "    final_star_allele = 'Ambiguous'\n",
    "    if math.isclose(best_site_percentage, 1.0):\n",
    "        final_star_alleles = set(best_site_stars).intersection(\n",
    "            best_genomic_stars\n",
    "        )\n",
    "        if len(final_star_alleles) == 1:\n",
    "            final_star_allele = list(final_star_alleles)[0]\n",
    "    # Has no sites- assign if best genomic match and no adequate site match\n",
    "    elif '*1.001' in best_genomic_stars:\n",
    "        final_star_allele = '*1.001'\n",
    "    return final_star_allele\n",
    "\n",
    "\n",
    "def get_best_site_matches(sites):\n",
    "    '''Get star alleles with highest and second-highest % matching and total\n",
    "    sites for hap, where second-highest is from different major star allele'''\n",
    "    # decreasing % then # of sites matched\n",
    "    sites = sorted(\n",
    "        sites, key=lambda x:(float(x.match) / x.total, x.total), reverse=True\n",
    "    )\n",
    "    best_percentage = match_frac(sites[0])\n",
    "    best_num_sites = sites[0].total\n",
    "    second_best_percentage = None\n",
    "    second_best_num_sites = None\n",
    "    best_site_matches = []\n",
    "    second_best_site_matches = []  # from different major star allele only\n",
    "    for site in sites:\n",
    "        if (\n",
    "            math.isclose(match_frac(site), best_percentage) and\n",
    "            site.total == best_num_sites\n",
    "        ):  # float equals\n",
    "            best_site_matches.append(site)\n",
    "        elif (\n",
    "            site.star_allele.split('.')[0] in\n",
    "            [z.star_allele.split('.')[0] for z in best_site_matches]\n",
    "        ):\n",
    "            continue\n",
    "        elif second_best_percentage is None:\n",
    "            second_best_percentage = match_frac(site)\n",
    "            second_best_site_matches.append(site)\n",
    "            second_best_num_sites = site.total\n",
    "        elif (\n",
    "            math.isclose(match_frac(site), second_best_percentage) and\n",
    "            site.total == second_best_num_sites\n",
    "        ):\n",
    "            second_best_site_matches.append(site)\n",
    "        else:\n",
    "            break\n",
    "    return (\n",
    "        best_percentage, best_site_matches, second_best_percentage,\n",
    "        second_best_site_matches\n",
    "    )\n",
    "\n",
    "\n",
    "def match_frac(site):\n",
    "    return float(site.match) / site.total\n",
    "\n",
    "\n",
    "def match_frac_str(site_list):\n",
    "    return ','.join([f'{site.match}/{site.total}' for site in site_list])\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get diplotypes\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "\n",
    "star_alleles_tsv = sys.argv[1]\n",
    "\n",
    "best_star_allele_per_hap = {}\n",
    "with open(star_alleles_tsv) as f:\n",
    "    f.readline()\n",
    "    for line in f:\n",
    "        line = line.strip().split('\\t')\n",
    "        hap = line[0]\n",
    "        best_star_allele = line[-1]\n",
    "        if best_star_allele != 'Ambiguous':\n",
    "            best_star_allele_per_hap[hap] = best_star_allele\n",
    "\n",
    "diplotypes = {}\n",
    "for hap in best_star_allele_per_hap:\n",
    "    sample = hap.split('_')[0]\n",
    "    if sample not in diplotypes:\n",
    "        diplotypes[sample] = [None, None]\n",
    "    star_allele_main = best_star_allele_per_hap[hap].split('.')[0]\n",
    "    if hap.endswith('_hap1'):\n",
    "        diplotypes[sample][0] = star_allele_main\n",
    "    else:\n",
    "        diplotypes[sample][1] = star_allele_main\n",
    "\n",
    "samples_per_diplotype = defaultdict(set)\n",
    "for sample in diplotypes:\n",
    "    hap1, hap2 = diplotypes[sample]\n",
    "    hap1_int = int(hap1.replace('*', '')) if (hap1) else None\n",
    "    hap2_int = int(hap2.replace('*', '')) if (hap2) else None\n",
    "    dip_strs = [\n",
    "        f\"{hap1 if (hap1) else '?'}\",\n",
    "        f\"{hap2 if (hap2) else '?'}\"\n",
    "    ]\n",
    "    # sort\n",
    "    dip_str = '/'.join(dip_strs[::-1]) if (hap2_int is not None and (hap1_int is None or hap2_int < hap1_int)) else '/'.join(dip_strs)\n",
    "    samples_per_diplotype[dip_str].add(sample)\n",
    "for x in sorted(samples_per_diplotype.items(), key=lambda z:len(z[1]), reverse=True):\n",
    "    print(f\"{x[0]}\\t{len(x[1])}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
